Add organization ID to client connection using header

medains opened the pull request:
This seems like the right place to do so - the resource API calls that require an organization ID that are not-quite RESTful also would need authentication with a user that is an administrator for the organization.  So putting it in at this level works out, and also avoids adding orgID parameters all over the place.

Co-authored-by: hansnqyr <luke@nqyr.io>

This change was previously submitted to https://github.com/nytm/go-grafana-api/pull/62
It is a prerequisite for this PR: https://github.com/grafana/terraform-provider-grafana/pull/110

aknuds1 reviewed (changes requested):
There are some changes I'd like to see, but first checking with my colleagues about the proposal.

aknuds1 reviewed (changes requested):
Just some quick feedback for now, to collect feedback until I have the time for a real review:

* Org ID should not be provided if authing with an API key, because an API key is for a single org and role, so only applicable with basic auth
* Making org ID mandatory for creating a client is a bad idea
* Allowing the org ID header to be set when using basic auth makes sense

medains committed:
Add organization ID to client connection using header.

This seems like the right place to do so - the resource API calls that
require an organization ID that are not-quite RESTful also would need
authentication with a user that is an administrator for the
organization.  So putting it in at this level works out, and also avoids
adding orgID parameters all over the place.

Co-authored-by: hansnqyr <luke@nqyr.io>

aknuds1 commented on client_test.go:
```diff
@@ -36,6 +36,23 @@ func TestNew_tokenAuth(t *testing.T) {
 	}
 }
 
+func TestNew_orgId(t *testing.T) {
```
```suggestion
func TestNew_orgID(t *testing.T) {
```

aknuds1 reviewed (changes requested):
Looks much better now, thanks! I will try to get feedback from @marefr this week to make sure this change makes sense.

Requesting changes to signal it still needs to be decided on.

aknuds1 reviewed (approved):
LGTM, but @marefr please review given your knowledge of the functionality.

aknuds1 committed:
Merge branch 'master' into master

marefr commented on client.go:
```diff
@@ -30,6 +30,8 @@ type Config struct {
 	// Client provides an optional HTTP client, otherwise a default will be used.
 	Client *http.Client
+	// OrgID provides an optional organization ID, ignored when using APIKey, BasicAuth defaults to last used org
+	OrgID string
```
I suggest this to be an `int64` instead to mimic the behavior in Grafana. If it's greater than zero it's supposed to be set.

medains replied:
Updated to int64, thanks

marefr reviewed (changes requested):
Minor change. See comment. Otherwise looks good I think :+1:

medains committed:
Change OrgID from string to int64

marefr reviewed (approved):
LGTM

aknuds1 committed:
Update client_test.go

gw0 commented on client.go:
```diff
@@ -103,6 +106,8 @@ func (c *Client) newRequest(method, requestPath string, query url.Values, body i
 
 	if c.config.APIKey != "" {
 		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", c.config.APIKey))
+	} else if c.config.OrgID != 0 {
```
I know this is already merged, but this line is wrong. This two cases are not mutually exclusive.

woodsaj replied:
You can't set an X-Grafana-Org-Id header when authenticating with an APIKey (well, you can set it, but it will be ignored).

APIKeys are already scoped to a specific org within Grafana.

aknuds1 replied:
@woodsaj Thanks, was about to write the same :)

gw0 replied:
Sorry for the noise, at first glimpse it seemed wrong.
